tests/cover/test_regex.py

```diff
@@ -0,0 +1,399 @@
+    u'ab{5}', u'ab{5,10}', u'ab{,10}', u'ab{5,}',  # repeaters
+    u'ab|cd|ef',  # branch
+    u'(foo)+', u'([\'"])[a-z]+\\1',
+    u'(?:[a-z])([\'"])[a-z]+\\1', u'(?P<foo>[\'"])[a-z]+(?P=foo)',  # groups
+    u'^abc',  # beginning
+    u'\d', u'[\d]', u'[^\D]', u'\w', u'[\w]', u'[^\W]',
+    u'\s', u'[\s]', u'[^\S]',  # categories
+])
+@pytest.mark.parametrize('encode', [False, True])
+def test_can_generate(pattern, encode):
+    if encode:
+        pattern = pattern.encode('ascii')
+    assert_all_examples(st.from_regex(pattern), re.compile(pattern).search)
+
+
+@pytest.mark.parametrize('pattern', [
+    re.compile(u'a', re.IGNORECASE),
+    u'(?i)a',
+    re.compile(u'[ab]', re.IGNORECASE),
+    u'(?i)[ab]',
+])
+def test_literals_with_ignorecase(pattern):
+    strategy = st.from_regex(pattern)
+
+    strategy.filter(lambda s: s == u'a').example()
+    strategy.filter(lambda s: s == u'A').example()
+
+
+@pytest.mark.parametrize('pattern', [
+    re.compile(u'[^a][^b]', re.IGNORECASE),
+    u'(?i)[^a][^b]'
+])
+def test_not_literal_with_ignorecase(pattern):
+    assert_all_examples(
+        st.from_regex(pattern),
+        lambda s: s[0] not in (u'a', u'A') and s[1] not in (u'b', u'B')
+    )
+
+
+def test_any_doesnt_generate_newline():
+    assert_all_examples(st.from_regex(u'.'), lambda s: s != u'\n')
+
+
+@pytest.mark.parametrize('pattern', [re.compile(u'.', re.DOTALL), u'(?s).'])
+def test_any_with_dotall_generate_newline(pattern):
+    st.from_regex(pattern).filter(lambda s: s == u'\n').example()
+
+
+@pytest.mark.parametrize('pattern', [re.compile(b'.', re.DOTALL), b'(?s).'])
+def test_any_with_dotall_generate_newline_binary(pattern):
+    st.from_regex(pattern).filter(lambda s: s == b'\n').example()
+
+
+@pytest.mark.parametrize('pattern', [
+    u'\d', u'[\d]', u'[^\D]',
+    u'\w', u'[\w]', u'[^\W]',
+    u'\s', u'[\s]', u'[^\S]',
+])
+@pytest.mark.parametrize('is_unicode', [False, True])
+@pytest.mark.parametrize('invert', [False, True])
+def test_groups(pattern, is_unicode, invert):
+    if u'd' in pattern.lower():
+        group_pred = is_digit
+    elif u'w' in pattern.lower():
+        group_pred = is_word
+    else:
+        # Special behaviour due to \x1c, INFORMATION SEPARATOR FOUR
+        group_pred = is_unicode_space if is_unicode else is_space
+
+    if invert:
+        pattern = pattern.swapcase()
+        _p = group_pred
+
+        def group_pred(s):
+            return not _p(s)
+
+    pattern = u'^%s\\Z' % (pattern,)
+
+    compiler = unicode_regex if is_unicode else ascii_regex
+    strategy = st.from_regex(compiler(pattern))
+
+    strategy.filter(group_pred).filter(is_ascii).example()
+    if is_unicode:
+        strategy.filter(lambda s: group_pred(s) and not is_ascii(s)).example()
+
+    assert_all_examples(strategy, group_pred)
+
+
+def test_caret_in_the_middle_does_not_generate_anything():
+    r = re.compile(u'a^b')
+
+    with pytest.raises(NoExamples):
+        st.from_regex(r).filter(r.search).example()
```

Comment:
IIUC, this doesn't test anything about `from_regex`.  This test would pass with `st.text()` because `r.search` will never return a match:
```pytb
>>> st.text().filter(re.compile(r'a^b').search).example()
Traceback (most recent call last):
  File "/Users/lucaswiman/opensource/hypothesis-python/.tox/py36/lib/python3.6/site-packages/hypothesis/searchstrategy/strategies.py", line 139, in example
    verbosity=Verbosity.quiet,
  File "/Users/lucaswiman/opensource/hypothesis-python/.tox/py36/lib/python3.6/site-packages/hypothesis/core.py", line 822, in find
    runner.valid_examples,))
hypothesis.errors.Unsatisfiable: Unable to satisfy assumptions of lambda x: True. Only 0 examples considered satisfied assumptions

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lucaswiman/opensource/hypothesis-python/.tox/py36/lib/python3.6/site-packages/hypothesis/searchstrategy/strategies.py", line 144, in example
    u'Could not find any valid examples in 100 tries'
hypothesis.errors.NoExamples: Could not find any valid examples in 100 tries
```